stackslib/src/net/httpcore.rs

```diff
@@ -0,0 +1,1625 @@
+                Ok(x) => x,
+                Err(_e) => {
+                    return Err(HttpError::Http(400, format!("Failed to decode `{}`", key)));
+                }
+            }
+        } else {
+            return Err(HttpError::Http(404, format!("Missing `{}`", key)));
+        };
+        Ok(u)
+    }
+}
+
+/// Extension to HttpRequestContents to give it awareness of Stacks-specific fields
+pub trait HttpRequestContentsExtensions {
+    /// Chain constructor: Request a specific tip
+    fn for_specific_tip(self, tip: StacksBlockId) -> Self;
+    /// Chain constructor: use a given TipRequest
+    fn for_tip(self, tip_req: TipRequest) -> Self;
+    /// Identify the tip request
+    fn tip_request(&self) -> TipRequest;
+    /// Determine if we should return a MARF proof
+    fn get_with_proof(&self) -> bool;
+}
+
+impl HttpRequestContentsExtensions for HttpRequestContents {
+    /// Request a specific tip
+    fn for_specific_tip(self, tip: StacksBlockId) -> Self {
+        self.query_arg("tip".to_string(), format!("{}", &tip))
+    }
+
+    /// Use a particular tip request
+    fn for_tip(mut self, tip_req: TipRequest) -> Self {
+        if tip_req != TipRequest::UseLatestAnchoredTip {
+            self.query_arg("tip".to_string(), format!("{}", &tip_req.to_string()))
+        } else {
+            let _ = self.take_query_arg(&"tip".to_string());
+            self
+        }
+    }
+
+    /// Ref the tip request
+    fn tip_request(&self) -> TipRequest {
+        self.get_query_args()
+            .get("tip")
+            .map(|tip| tip.as_str().into())
+            .unwrap_or(TipRequest::UseLatestAnchoredTip)
+    }
+
+    /// Get the proof= query parameter value
+    fn get_with_proof(&self) -> bool {
+        let with_proof = if let Some(proof_val) = self.get_query_arg(&"proof".to_string()) {
+            proof_val == "1"
+        } else {
+            false
+        };
+
+        with_proof
+    }
+}
+
+/// Work around Clone blanket implementations not being object-safe
+pub trait RPCRequestHandlerClone {
+    fn clone_rpc_handler_box(&self) -> Box<dyn RPCRequestHandler>;
+}
+
+impl<T> RPCRequestHandlerClone for T
+where
+    T: 'static + RPCRequestHandler + Clone,
+{
+    fn clone_rpc_handler_box(&self) -> Box<dyn RPCRequestHandler> {
+        Box::new(self.clone())
+    }
+}
+
+impl Clone for Box<dyn RPCRequestHandler> {
+    fn clone(&self) -> Box<dyn RPCRequestHandler> {
+        self.clone_rpc_handler_box()
+    }
+}
+
+/// Trait that every HTTP round-trip request type must implement.
+pub trait RPCRequestHandler: HttpRequest + HttpResponse + RPCRequestHandlerClone {
+    /// Reset the RPC handler.  This clears any internal state this handler stored between calls to
+    /// `try_handle_request()`
+    fn restart(&mut self);
+    /// Instantiate the HTTP response headers and body from a request
+    fn try_handle_request(
+        &mut self,
+        request_preamble: HttpRequestPreamble,
```

Comment:
I should really comment on the trait...less of the functions... :P So I think you may be able to take references for the request_preamble and request_body....Though I didn't go through every single implementation...

Reply:
The reason the `RPCRequestHandler::try_handle_request()` function implementation calls for a pass-by-value instead of a pass-by-reference is to avoid requiring the caller to perform extra allocations to get owned copies of parts of the request headers and body.  I don't think it hurts anything to keep it as pass-by-value, since the caller of `.try_handle_request()` does not need to make copies either.